src/Microsoft.OData.Client/BatchSaveResult.cs

```diff
@@ -370,12 +370,27 @@ private ODataRequestMessageWrapper GenerateBatchRequest()
 
                     for (int i = 0; i < this.ChangedEntries.Count; ++i)
                     {
+                        Descriptor descriptor = this.ChangedEntries[i];
+
                         if (Util.IsBatchWithIndependentOperations(this.Options))
                         {
-                            this.batchWriter.WriteStartChangeset();
+                            if(descriptor.ChangeSetId != null)
+                            {
+                                this.batchWriter.WriteStartChangeset(descriptor.ChangeSetId);
+                            }
+                            else
+                            {
+                                this.batchWriter.WriteStartChangeset();
+                            }
+                        }
+
+                        // In BatchWithIndependentOperations, we use atomicityGroup(in this case changeSetId)
+                        // as the dependsOn property
+                        if(descriptor.DependsOnIds != null && Util.IsBatchWithIndependentOperations(this.Options))

```

Comment:
@KenitoInc Since the `if` block above this one also calls `Util.IsBatchWithIndependentOperations(this.Options)`, you could store the result in a variable declared in the same scope as the `descriptor` variable.

Reply:
Resolved